React + Snowpack Guide (#1309)
snowpackjs/snowpack: www/guides/react.md

```
📁 src : your React components and their assets (CSS, images)
↳ index.jsx
📁 public : global assets like images, fonts, icons, and global CSS
↳ index.css
↳ index.html
```

Use your favorite visual editor to rearrange and rename, or run these commands in the terminal:

```bash
mkdir src
mkdir public
mv index.jsx src/index.jsx
mv index.html public/index.html
mv index.css public/index.css
```

This means if you are running Snowpack right now, the site is now broken as the files are all in different places.

Snowpack Configuration allows you to fix this and by changing what directories Snowpack scans and where they output. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a configuration file with empty options. Add this to the empty `mount` object:

```diff
  mount: {
-   /* ... */
+   // directory name: 'build directory'
+   public: '/',
+   src: '/_dist_',
  },
```

> <img src="/img/guides/react/folderstructure.jpg" alt="The original file configuration had Snowpack building the directory structure the same as the directories in the project, including root. Now the config builds only src and public. Src to the dist folder and public to root." />

Author: yeah this is much better!

  color: white;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
```

To use this CSS, head to `App.jsx` and import it

```diff
  import logo from './logo.svg';
+ import './App.css';
```

<div class="frame"><img src="/img/guides/react/react.gif" alt="The page now has centered items, a grey background, styled fonts, and the React logo has an animation that rotates it." class="screenshot"/></div>

## Making Snowpack Even Faster with Fast Refresh

[React Fast Refresh](https://reactnative.dev/docs/fast-refresh)? What's that? It's a developer tool that makes it much easier to see how updates to your code work in the browser. React projects are often interactive and include state. For example, this project you're building has a state that is the amount of time on the page. When developing with state it's useful not to lose it while you edit code. React Fast Refresh shows you updates without refreshing the entire page. Showing you how to add this is also a good intro to Snowpack plugins. Snowpack starts with a minimal setup with the perspective that you can add what you need through the plugin system.

> Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement), a requirement for Fast Refresh. Snowpack automatically refreshes to show changes in your code. Hot Module Replacement only updates the components that changed without refreshing the whole page. You can enable it for React with a small snippet of code in `index.jsx`.

Reviewer: clarify the difference: 

> Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement) in your project. HMR is the system that lets Snowpack push updates to the browser without a full page refresh,  a requirement for Fast Refresh. You can enable HMR for React by adding a small snippet of code to your `src/index.jsx` file.

> Head to `index.jsx` and add this snippet at the bottom:

Reviewer: Don't think this line is needed with the new edit on the paragraph above (see comment above)

```diff
  ReactDOM.render(
    <React.StrictMode>
      <App />
    </React.StrictMode>,
    document.getElementById('root'),
  );
+ // Hot Module Replacement (HMR) - Remove this snippet to remove HMR.
+ // Learn more: https://www.snowpack.dev/#hot-module-replacement
+ if (import.meta.hot) {
+   import.meta.hot.accept();
+ }
```

> Now when you change `App.jsx` the changes show without the whole page refreshing.

Reviewer: "Now when you change `App.jsx` the page updates to show your changes without a full refresh."

<div class="frame"><img src="/img/guides/react/hmr.gif" alt="GIF showing code side by side with the app. A change in made to App.jsx and it shows immediately when the file is changed. The counter keeps counting uninterrupted." class="screenshot"/></div>

> The counter on the page still resets to 0, even though the page itself hasn't fully refreshed. This can be annoying if you're trying to debug state. React Fast Refresh solves this.

Reviewer: "HMR can save you time on its own, but you may notice in the example above that the counter on the page still resets to 0. This can slow down your development, especially when you're trying to debug a specific component state problem. Lets enable Fast Refresh to preserve component state across updates.

To enable Fast Refresh, we'll need to install the "@snowpack/plugin-react-refresh" package. This package is a Snowpack plugin, which you can use to enhance or customize Snowpack with all sorts of new behaviors. To start, install the package in your project:

Reviewer: (this comment is can replace line 305-309)

Installing a plugin involves installing the module and then enabling it in the Snowpack configuration file.

Install the module first by running.

```bash
npm install @snowpack/plugin-react-refresh --save-dev
```

> To enable it head to `snowpack-config.js` and add it to the array of plugins

Reviewer: "Once installed, you'll need to add the plugin to your Snowpack configuration file so that Snowpack knows to use it".

```diff
  module.exports = {
    mount: {
      public: '/',
      src: '/_dist_',
    },
-   plugins: []
+   plugins: ['@snowpack/plugin-react-refresh'],
  };
```

> > 💡 Tip: you'll need to restart Snowpack for the configuration changes to take effect

Reviewer: instead of having this in a blockquote, start the next paragraph with this instruction. Blockquotes are only used above as additional info and not important/required steps.

> Now make an edit to `App.jsx`. If this worked you'll see it without the counter stopping or losing its value.

Reviewer: "Restart Snowpack to apply the new plugin, and then try changing the `App.jsx`  component again. If Fast Refresh is working properly you ill see the counter keep its value across changes, without resetting to zero.

<div class="frame"><img src="/img/guides/react/react-fast-refresh.gif" alt="GIF showing code side by side with the app. A change in made to App.jsx and it shows immediately when the file is changed. The counter keeps counting uninterrupted." class="screenshot"/></div>

## Going further

> Great job! You're now ready to build the React project of your dreams with Snowpack.

Reviewer: Good to call out that this is the end. "Congratulations! You've reached the end of our guide, and are now ready to..."

---
layout: layouts/guide.njk
permalink: '/guides/react/'
title: The Snowpack Guide to Getting Started with React
description: 'Get started with this in-depth tutorial on how to build React applications and websites with Snowpack. '
date: 2020-10-01
---

Image: ideally a nice header image showing the before or after, or some illustration with elements of the Snowpack logo, React, and symbols related to speed, joy.

> Snowpack is a great fit for [React](https://reactjs.org/) projects of any size. It's easy to get started with and can scale to projects containing thousands of components and pages without any impact on development speed. Unlike traditional React application tooling, Snowpack saves you from getting bogged down with complex bundler setups and configuration files.

Reviewer: It's easy to get started ~~with~~ and can scale?

> In this guide, you'll go from an empty directory to a fully configured Snowpack project with support for React and several other useful optimizations. In the process, you'll learn:

Reviewer: ~optimizations~ features?

- How to set up Snowpack
- How to get started with React in a Snowpack project
- Importing CSS, images and other assets
- Enabling [Fast Refresh](https://reactnative.dev/docs/fast-refresh) mode for React
- Connecting your favorite developer tools to Snowpack

> 💡 Tip: if you want to jump to the end to see a full featured React setup, the [Create Snowpack App React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) comes with everything you'll learn in this tutorial plus other useful tools.
>
> ```bash
> npx create-snowpack-project react-snowpack-template --template @snowpack/project-template-react
> ```

Prerequisites: Snowpack is a command line tool installed from npm. This guide assumes a basic understanding of Node.js, npm, and how to run commands in the terminal. Knowledge of React is not required, Snowpack is a great way to learn React!

## Getting started with Snowpack and React

The easiest way to start a new Snowpack project is with [Create Snowpack App](<https://www.snowpack.dev/#create-snowpack-app-(csa)>). Create Snowpack App is a tool to quickly set up Snowpack in a new directory.

> The template you're starting with here is `@snowpack/project-template-minimal` is a bare-bones setup where Snowpack serves an `index.html`, `index.css`, and `index.js`.

Reviewer: The template that you will start with here is

or 

In this guide, you will use the `@snowpack/project-template-minimal` template to get started with a simple, bare-bones Snowpack project setup. The rest of the guide will build on this.

> Open up your terminal and head to a directory where you want to put your new project. Now run `npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal`. This creates a new directory called `react-snowpack` and installs the `@snowpack/project-template-minimal` inside it.

Reviewer: Start with "To get started, open your terminal..."?

> You can now head to that new `react-snowpack` directory with `cd react-snowpack`. And to start up Snowpack run `npm run start`. You should see your new website up and running!

Reviewer: Here and above: all commands that the user should run should look the same, stylistically. That means moving 

```
npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal
```

and 

```
cd react-snowpack
npm run start
```

into their own full code blocks. 

The text you have don't necessarily have to change,  although the create-snowpack-app command is long enough that you could probably replace that text with:

> Now run the following command in your terminal to create a new directory called `react-snowpack` that will automatically be filled with the minimal template & Snowpack setup.

> 💡 Tip: the `README.md` in your new project contains useful information about what each file does.

Image: screenshot of project-template-minimal, which shows "Hello world" in text on a white background.

> 💡 Tip: you can also use Yarn for any of the commands in this tutorial.

> Now that you have a basic project up and running, it's time to install React. To install React, run the following command in your project directory:

Reviewer: install twice, maybe "it's time to add React"?

```bash
npm install react react-dom --save
```

> 💡 Tip: in this tutorial commands use `--save` for frontend libraries and `--save-dev` for your developer tools to keep them organized separately. This is totally up to you, though.

## Create your first React component

> Snowpack has built in support for JSX in files with the `.jsx` extension. That means that there's no additional plugins or configuration needed to write your first React component, but you will need to rename your `index.js` file to `index.jsx` to take advantage of this.

Reviewer: New user: "what's JSX?"

Prefix this paragraph with "React relies on a special templating language called JSX. Snowpack has..."

```bash
mv index.js index.jsx
```

> 💡 Tip: You do not need to update your `index.html` script tag to point to `index.jsx`. Because JSX isn't supported in the browser, Snowpack builds your file to a runnable `.js` file. This is good to remember when you're referencing built files in HTML `<script src="">` and `<link href="">` elements.

You can now import React in `index.jsx` and add a simple test component just to make sure it's working:

```diff
/* Add JavaScript code here! */
- console.log('Hello World! You did! Welcome to Snowpack :D');
+import React from 'react';
+import ReactDOM from 'react-dom';
+ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));
```

Since the React code is rendering into an element with the ID `root`, we'll need to add that to `index.html`:

```diff
  <body>
-    <h1>Welcome to Snowpack!</h1>
+    <div id="root"></div>
    <script type="module" src="/index.js"></script>
```

Image: screenshot of the project, which shows "HELLO REACT" on a white background

> You've just created your first React component in Snowpack! And also gotten your project prepared for building more with React.

Reviewer: bad grammar with "And also", i think that second sentence is unnecessary.

> ## Customizing your file layout

Reviewer: ~~Customizing~~ Customize

Also "project layout" instead of file layout?

> If you're going to be adding a bunch of new files, you probably don't want them in the root directory. Snowpack is flexible enough to support whatever project layout that you prefer. For this guide, let's use a popular project layout from the React community.

Reviewer: New user: Why not?

maybe instead: "you probably don't want them crowding up your top-level root directly."

> If you're going to be adding a bunch of new files, you probably don't want them in the root directory. Snowpack is flexible enough to support whatever project layout that you prefer. For this guide, let's use a popular project layout from the React community.

Reviewer: Something about "If you're going to be adding" is also reading funny. Maybe more "Since you'll be adding"?

A standard way of organizing React projects is an `src` directory for components and their associated files. Then a `public` directory for static assets like images, fonts, icons, and sometimes CSS. Lets update our project to use this popular structure.

```
📁 src
↳ index.jsx
📁 public
↳ index.css
↳ index.html
```

Provided are some commands that give the above project structure, you may prefer to use other tools like an IDE or text editor for this:

```bash
mkdir src
mkdir public
mv index.jsx src/index.jsx
mv index.html public/index.html
mv index.css public/index.css
```

> Once your new structure is in place, you need to tell Snowpack about it. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a mostly-empty, bare-bones config file. Add this to the empty `mount` object:

Reviewer: "Once your new file structure" or "Once your new project structure"

## Styling your Snowpack/React project

When you add assets like images or CSS, Snowpack includes them in your final build. If you already know React, this process should look pretty familiar.

> 💡 Tip: As you're doing this, you should not need to reload the page or restart Snowpack. Snowpack automatically updates the project in the browser as you edit code.

Add this file `logo.svg` (link to file) to your `src` directory. Now you can import it into your `App.jsx` and use it in an `img` tag to display it.

```diff
import React, { useState, useEffect } from 'react';
+ import logo from './logo.svg';

function App() {
  // Create the count state.
  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);
  }, [count, setCount]);
  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">

> +        <img src={logo} className="App-logo" alt="logo" />

Reviewer: can you add the 1 line below this one, so that people get a better idea of where this should go in their file?

```bash
npm install react react-dom --save
```

> 💡 Tip: in this guide commands use `--save` for frontend libraries and `--save-dev` for your developer tools to keep them organized separately. This is totally up to you, though.

## Create your first React component

React relies on a special templating language called JSX. Snowpack has built in support for JSX in files with the `.jsx` extension. That means that there's no plugins or configuration needed to write your first React component. Rename `index.js` file to `index.jsx` so that Snowpack knows it's now a JSX file:

```bash
mv index.js index.jsx
```

> 💡 Tip: you do not need to update your `index.html` script tag to point to `index.jsx`. Because JSX isn't supported in the browser, Snowpack builds your file to a runnable `.js` file. This is good to remember when you're referencing built files in HTML `<script src="">` and `<link href="">` elements.

You can now import React in `index.jsx` and add a simple test component just to make sure it's working:

```diff
/* Add JavaScript code here! */
- console.log('Hello World! You did! Welcome to Snowpack :D');
+import React from 'react';
+import ReactDOM from 'react-dom';

> +ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));

Reviewer 2: ```suggestion
- console.log('Hello World! You did! Welcome to Snowpack :D');
+ import React from 'react';
+ import ReactDOM from 'react-dom';
+ ReactDOM.render(<div>"HELLO REACT"</div>, 
  document.getElementById('root'));
```

```

Since the React code is rendering into an element with the ID `root`, you'll need to add that to `index.html`:

```diff
  <body>
-    <h1>Welcome to Snowpack!</h1>

> +    <div id="root"></div>

Reviewer 2: ```suggestion
  <body>
-   <h1>Welcome to Snowpack!</h1>
+   <div id="root"></div>
+   <script type="module" src="/index.js"></script>
  </body>
```

    <script type="module" src="/index.js"></script>
```

<div class="frame"><img src="/img/guides/react/minimalist-hello-world-react.png" alt="screenshot of the project, which shows 'HELLO REACT' on a white background" class="screenshot"/></div>

You've just created your first React component in Snowpack!

## Customize your project layout

Since you'll be adding a bunch of new files, you probably don't want them crowding up your top-level root directly. Snowpack is flexible enough to support whatever project layout that you prefer. In this guide, you'll learn how to use a popular project pattern from the React community.

```
📁 src : your React components and their assets (CSS, images)
↳ index.jsx
📁 public : global assets like images, fonts, icons, and global CSS
↳ index.css
↳ index.html

> ```

Reviewer 2: Also, see my below comment, but now that I’m looking at this I think a screenshot of the VS Code sidebar may be easier to recognize? That could establish familiarity with what developers look at

Use your favorite visual editor to rearrange and rename, or run these commands in the terminal:

```bash
mkdir src
mkdir public
mv index.jsx src/index.jsx
mv index.html public/index.html
mv index.css public/index.css
```

This changes how Snowpack builds the files since by default Snowpack scans all directories:

```
/ (builds to → /)
📁 src
↳ index.jsx (builds to → src/index.js)
📁 public
↳ index.css (builds to → public/index.css)
↳ index.html (builds to → public/index.html)

> ```

Reviewer 2: ```suggestion
├── 📁 src/
│   └── index.jsx → src/index.js
└── 📁 public/
    ├── index.css → public/index.css
    └── index.html → public/index.html
```
Reading this again in the guide with everything else going on, it was hard for me to see this as a tree structure. I think using the “box drawing” unicode characters, and the convention of two dashes per file (`──`) make the file structure quicker to recognize at-a-glance

Reviewer 2: Also, I think you could just use `→` in the tree structure, and add a note below like “→ represents the file that Snowpack will build (note in some cases Snowpack will overwrite what’s there)”

```diff
  mount: {
-    /* ... */
+ // directory name: 'build directory'
+    public: '/',
+    src: '/_dist_',
  },
```

This configuration changes the build to:

```
/ (ignored in case you have files in here you might not want in your final build)
📁 src
↳ index.jsx (builds to → _dist_/index.js)
📁 public
↳ index.css (builds to → index.css)
↳ index.html (builds to → index.html)
```

`mount` is part of the [Snowpack Configuration API](https://www.snowpack.dev/#configuration). It allows you to customize the file structure of your project. The key is the name of the directory and the value is where you'd like them in the final build. With this new configuration, Snowpack builds files in `public` like `public/index.css` directory into `index.css`. It builds files in `src` like `src/index.js` into `/_dist_/index.js`, so you'll need to change that path in your `index.html`:

```diff
    <h1>Welcome to Snowpack!</h1>
    <div id="root"></div>
-    <script type="module" src="/index.js"></>

> +    <script type="module" src="/_dist_/index.js"></script>

Reviewer 2: ```suggestion
  <body>
    <h1>Welcome to Snowpack!</h1>
    <div id="root"></div>
-   <script type="module" src="/index.js"></script>
+   <script type="module" src="/_dist_/index.js"></script>
  </body>
```

  </body>
```

You'll need to restart Snowpack for configuration file changes. When you start up again, if it worked, it should look the same.

Create a new file at `src/App.jsx` and paste the following code into this new file to create an `App` component:

```jsx
import React, {useState, useEffect} from 'react';

function App() {
  // Create the count state.
  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);
  }, [count, setCount]);
  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">
        <p>
          Page has been open for <code>{count}</code> seconds.
        </p>
      </header>
    </div>
  );
}

export default App;
```

Now include it in `index.jsx`

```diff
import React from 'react';
import ReactDOM from 'react-dom';
+ import App from './App.jsx';

- ReactDOM.render(<div>"HELLO WORLD"</div>, document.getElementById('root'));
+ ReactDOM.render(
+    <React.StrictMode>
+      <App />
+    </React.StrictMode>,
+    document.getElementById('root'),

> +  );

Reviewer 2: ```suggestion
  import React from 'react';
  import ReactDOM from 'react-dom';
+ import App from './App.jsx';
- ReactDOM.render(<div>"HELLO WORLD"</div>, document.getElementById('root'));
+ ReactDOM.render(
+   <React.StrictMode>
+     <App />
+   </React.StrictMode>,
+   document.getElementById('root'),
+ );
```

## Styling your Snowpack/React project

When you add assets like images or CSS, Snowpack includes them in your final build. If you already know React, this process should look pretty familiar.

> 💡 Tip: as you're doing this, you should not need to reload the page or restart Snowpack. Snowpack automatically updates the project in the browser as you edit code.

Add this file [`logo.svg`](https://github.com/snowpackjs/snowpack/blob/master/create-snowpack-app/app-template-react/src/logo.svg) to your `src` directory. Now you can import it into your `App.jsx` and use it in an `img` tag to display it.

```diff
import React, { useState, useEffect } from 'react';
+ import logo from './logo.svg';

function App() {
  // Create the count state.
  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);
  }, [count, setCount]);
  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">
+        <img src={logo} className="App-logo" alt="logo" />

>         <p>

Reviewer 2: ```suggestion
  function App() {
    // Create the count state.
    const [count, setCount] = useState(0);
    // Create the counter (+1 every second).
    useEffect(() => {
      const timer = setTimeout(() => setCount(count + 1), 1000);
      return () => clearTimeout(timer);
    }, [count, setCount]);
    // Return the App component.
    return (
      <div className="App">
        <header className="App-header">
+       <img src={logo} className="App-logo" alt="logo" />
        <p>
```

The project already has index.css for global styles. For CSS that's only for a specific component, a common design pattern is to add it in a CSS file with the same base name as the component. The style file for `App.jsx` would be `App.css` with this pattern.

> 💡 Tip: Snowpack has built-in support for [CSS Modules](https://www.snowpack.dev/#import-css-modules) and if you'd like to use SASS there is an official [SASS Plugin](https://www.snowpack.dev/#sass).

Create `src/App.css` and add this CSS:

```css
.App {
  text-align: center;
}

.App p {
  margin: 0.4rem;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
```

To use it, import it `App.jsx` with

```diff
import logo from './logo.svg';

> +import './App.css';

Reviewer 2: ```suggestion
  import logo from './logo.svg';
+ import './App.css';
```

Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement), a requirement for Fast Refresh. Snowpack automatically refreshes to show changes in your code. Hot Module Replacement only updates the components that changed without refreshing the whole page. You can enable it for React with a small snippet of code in `index.jsx`.

Head to `index.jsx` and add this snippet at the bottom:

```diff
ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root'),
);
+// Hot Module Replacement (HMR) - Remove this snippet to remove HMR.
+// Learn more: https://www.snowpack.dev/#hot-module-replacement
+if (import.meta.hot) {
+  import.meta.hot.accept();

> +}

Reviewer 2: ```suggestion
  ReactDOM.render(
    <React.StrictMode>
      <App />
    </React.StrictMode>,
    document.getElementById('root'),
  );
+ // Hot Module Replacement (HMR) - Remove this snippet to remove HMR.
+ // Learn more: https://www.snowpack.dev/#hot-module-replacement
+ if (import.meta.hot) {
+   import.meta.hot.accept();
+ }
```

```

Now when you change `App.jsx` the changes show without the whole page refreshing.

<div class="frame"><img src="/img/guides/react/hmr.gif" alt="GIF showing code side by side with the project. A change in made to App.jsx and it shows when the changed file is saved. But the counter resets to 0." class="screenshot"/></div>

The counter on the page still resets to 0, even though the page itself hasn't fully refreshed. This can be annoying if you're trying to debug state. React Fast Refresh solves this.

Installing a plugin involves installing the module and then enabling it in the Snowpack configuration file.

Install the module first by running.

```bash
npm install @snowpack/plugin-react-refresh --save-dev
```

To enable it head to `snowpack-config.js` and add it to the array of plugins

```diff
module.exports = {
  mount: {
    public: '/',
    src: '/_dist_',
  },
  plugins: [
+    '@snowpack/plugin-react-refresh'
  ],

> };

Reviewer 2: ```suggestion
  module.exports = {
    mount: {
      public: '/',
      src: '/_dist_',
    },
-   plugins: []
+   plugins: ['@snowpack/plugin-react-refresh'],
  };
```

Snowpack is a great fit for [React](https://reactjs.org/) projects of any size. It's easy to get started and can scale to projects containing thousands of components and pages without any impact on development speed. Unlike traditional React application tooling, Snowpack saves you from getting bogged down with complex bundler setups and configuration files.

In this guide, you'll go from an empty directory to a fully configured Snowpack project with support for React and several other useful developer tools. In the process, you'll learn:

- How to set up Snowpack
- How to get started with React in a Snowpack project
- Importing CSS, images and other assets
- Enabling [Fast Refresh](https://reactnative.dev/docs/fast-refresh) mode for React
- Connecting your favorite developer tools to Snowpack

> 💡 Tip: if you want to jump to the end to see a full featured React setup, the [Create Snowpack App React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) comes with everything you'll learn in this guide plus other useful tools.
>
> ```bash

> > npx create-snowpack-project react-snowpack-template --template @snowpack/project-template-react

Reviewer 2: Is there a reason this template name is different from the snippet right below? I don’t want super beginners to be confused something isn’t working if they copied this one by mistake

Reviewer 2: Oh I’m a dummy this is the finished project. WDYT about moving this snippet to the very end, and only linking to the last heading here? That way the only snippets they see are copy-safe (but can still jump to the end if they want)

```

Since the React code is rendering into an element with the ID `root`, you'll need to add that to `index.html`:

```diff
  <body>
-    <h1>Welcome to Snowpack!</h1>
+    <div id="root"></div>
    <script type="module" src="/index.js"></script>
```

Image: screenshot of the project, which shows "HELLO REACT" on a white background

You've just created your first React component in Snowpack!

## Customize your project layout

Since you'll be adding a bunch of new files, you probably don't want them crowding up your top-level root directly. Snowpack is flexible enough to support whatever project layout that you prefer. In this guide, you'll learn how to use a popular project pattern from the React community.

In this pattern the `src` directory is for components and their associated files. Then a `public` directory is for static assets like images, fonts, icons, and sometimes CSS.

```bash
📁 src
↳ index.jsx
📁 public
↳ index.css
↳ index.html
```

Use your favorite visual editor to rearrange and rename, or run these commands in the terminal:

```bash
mkdir src
mkdir public
mv index.jsx src/index.jsx
mv index.html public/index.html
mv index.css public/index.css
```

> Once your new project structure is in place, you need to tell Snowpack about it. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a mostly empty, bare-bones configuration file. Add this to the empty `mount` object:

Reviewer 2: I like this explainer section, but I think as a beginner I’d be confused why `src` and why `public`. What is the purpose of each?

Also by default, Snowpack scans the entire folder; `mount` just limits the scope. So it may be more accurate now to say “we may have files we don’t want to publish online, so let’s tell Snowpack to only build specific folders.” You may also want to mention this can be a speed boost, if we’re only scanning part of the project.

```diff
    <h1>Welcome to Snowpack!</h1>
    <div id="root"></div>
-    <script type="module" src="/index.js"></>
+    <script type="module" src="/_dist_/index.js"></script>
  </body>
```

If Snowpack is still running, exit the process, you need to restart Snowpack for configuration file changes. When you start up again, if it worked, it should look the same.

Create a new file at `src/App.jsx` and paste the following code into this new file to create an `App` component:

```jsx
import React, {useState, useEffect} from 'react';

function App() {
  // Create the count state.
  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);

>   }, [count, setCount]);

Reviewer 2: We don’t need `setCount` in the memo here (but also it’s not hurting anything)

Reviewer: some linters get mad if you don't :/ agreed it's not very important but good to do the right thing here in case anyone has one turned on by default

Author: that's just copy paste from the React template so maybe change it there?

Reviewer 2: > some linters get mad if you don't :/

Yeah I always disabled those because they’re wrong! 😛  Queue another discussion about React hooks. (I’m fine for keeping, because again here it doesn’t hurt anything)

  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">
        <p>
          Edit <code>src/App.jsx</code> and save to reload.
        </p>
        <p>
          Page has been open for <code>{count}</code> seconds.
        </p>
        <p>
          <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
        </p>
      </header>
    </div>
  );
}

export default App;
```

Now include it in `index.jsx`

```diff
import React from 'react';
import ReactDOM from 'react-dom';
+ import App from './App.jsx';

- ReactDOM.render(<div>"HELLO WORLD"</div>, document.getElementById('root'));
+ ReactDOM.render(
+    <React.StrictMode>
+      <App />
+    </React.StrictMode>,
+    document.getElementById('root'),
+  );
```

> > 💡 Tip: [Strict Mode](https://reactjs.org/docs/strict-mode.html) is a tool for highlighting potential problems in React code.

Reviewer 2: Love these little tips!

  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);
  }, [count, setCount]);
  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">
+        <img src={logo} className="App-logo" alt="logo" />
        <p>
```

Image: the React logo (a blue atom) is now at the top of the page

The project already has index.css for global styles. For CSS that's only for a specific component, a common design pattern is to add it in a CSS file with the same base name as the component. So the file for `App.jsx` would be `App.css` with this pattern.

> 💡 Tip: Snowpack has built-in support for [CSS Modules](https://www.snowpack.dev/#import-css-modules) and if you'd like to use SASS there is an official [SASS Plugin](https://www.snowpack.dev/#sass).

Create `src/App.css` and add this CSS:

```css
.App {
  text-align: center;
}
.App code {
  background: #fff3;
  padding: 4px 8px;
  border-radius: 4px;
}
.App p {
  margin: 0.4rem;
}
.App-logo {
  height: 40vmin;
  pointer-events: none;
}

> @media (prefers-reduced-motion: no-preference) {

Reviewer 2: ❤️ ❤️ ❤️

  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {

>     transform: rotate(0deg);

Reviewer 2: This affects nothing at all, but you can omit the `from [` block here as it’s inferred, to save lines

Image: GIF showing code side by side with the project. A change in made to App.jsx and it shows when the changed file is saved. But the counter resets to 0.

The counter on the page will still reset to 0, even though the page itself hasn't fully refreshed. This can be annoying if you're trying to debug state. React Fast Refresh solves this.

Installing a plugin involves installing the module and then enabling it in the Snowpack configuration file.

Install the module first by running.

```bash
npm install @snowpack/plugin-react-refresh --save-dev
```

To enable it head to `snowpack-config.js` and add it to the array of plugins

```diff
module.exports = {
  mount: {
    public: '/',
    src: '/_dist_',
  },
  plugins: [
+    '@snowpack/plugin-react-refresh'
  ],
};
```

> 💡 Tip: you'll need to restart Snowpack for the configuration changes to take effect

Now make an edit to `App.jsx`. If this worked you'll see it without the counter stopping or losing its value.

Image: GIF showing code side by side with the app. A change in made to App.jsx and it shows immediately when the file is changed. The counter keeps counting uninterrupted.

## Going further

> Great job! You're now ready to build the React project if your dreams with Snowpack.

Reviewer 2: ❤️

Reviewer 2: (also sp: `of`0

Image: Certificate with share buttons for Twitter

> At this point you the basics and have a great starter for any React project. But if you compare with the official [Snowpack React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) you'll notice it has some other developer tools you might find useful:

Reviewer 2: ```suggestion
At this point you have the basics and have a great starter for any React project. But if you compare with the official [Snowpack React template](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react) you'll notice it has some other developer tools you might find useful:
```

Author: you the basics lol

- [Prettier](https://prettier.io/) — a code formatter

- Tests- TODO

- [`@snowpack/plugin-dotenv`](https://github.com/pikapkg/snowpack/tree/master/plugins/plugin-dotenv) - Use `dotenv` in your Snowpack. This is useful for environment specific variables

> If you'd like to use Typescript with Snowpack and React, checkout the [Snowpack React Typescript](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react-typescript) starter.

Reviewer 2: ```suggestion
If you'd like to use TypeScript with Snowpack and React, check out the [Snowpack React TypeScript](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-react-typescript) starter.
```

> If you have any questions, comments, or corrections, we'd love to here from you in the Snowpack [discussion](https://github.com/pikapkg/snowpack/discussions) forum or our [Snowpack Discord community](https://discord.gg/rS8SnRk).

Reviewer 2: ```suggestion
If you have any questions, comments, or corrections, we'd love to hear from you in the Snowpack [discussion](https://github.com/pikapkg/snowpack/discussions) forum or our [Snowpack Discord community](https://discord.gg/rS8SnRk).
```

> ---

Reviewer 3: (Don't know where to put this) when I was searching for the guide via [https://snowpack-git-react-guide.pikapkg.vercel.app/](https://snowpack-git-react-guide.pikapkg.vercel.app/), finding the guide was more difficult than I expected. Perhaps there could be a list of guides on the sidebar on the main page, which would link to the appropriate guide?

layout: layouts/guide.njk
permalink: '/guides/react/'
title: Snowpack + React Guide
description: 'Documentation and guides for building projects with Snowpack and React.'
date: 2020-10-01
---

> Image: ideally a nice header image showing the before or after, or some illustration with elements of the Snowpack logo, React, and symbols related to speed, joy.

Reviewer 2: Nope, SGTM!

Snowpack is a great fit for [React](https://reactjs.org/) projects of any size. It's easy to get started with and can scale to projects containing thousands of components and pages without any impact on development speed. Unlike traditional React application tooling, Snowpack saves you from getting bogged down with complex bundler setups and configuration files.

> In this guide, you'll go from a [minimal Snowpack project](https://github.com/pikapkg/snowpack/tree/master/create-snowpack-app/app-template-minimal) to a Snowpack project with React and several useful developer optimizations. In the process you'll learn:

Reviewer: "In this guide, you'll go from an empty directory to a fully configured Snowpack project with support for React and several other useful optimizations. In the process, you'll learn:"

I think the minimal template is just step one of our template, i wouldn't want devs thinking they need to set this up themselves.

- How to set up a new Snowpack project

> - How to get started with React in a Snowpack project

Reviewer: ~~in a Snowpack project~~ is this implied? I think safe to remove

Reviewer: Actually, I think this would be better as "Adding your first React component"

> - Importing and using CSS and other assets inside your project

Reviewer: run-on sentence. How about "Importing CSS, Sass, Images and other assets"

(this would work if we had a mention for Sass somewhere in the CSS section. We get so many people asking about Sass, I think it makes sense to have a mention).

- Enabling [Fast Refresh](https://reactnative.dev/docs/fast-refresh) mode for React

> - Extending Snowpack with plugins

Reviewer: Maybe more "connecting your favorite tools to Snowpack" or "connecting your favorite dev tools"

> - Extending Snowpack with plugins

Reviewer 3: Perhaps there could be a link at the end of each of these, linking them to their appropriate section? I found myself wanting to go to the "Fast Refresh" section but there wasn't a quick link to go to it.

> > If you want to jump to the end to see a full featured React setup, the Create Snowpack App React template comes with everything you'll learn in this tutorial plus other useful tools.

Reviewer: add link here

> > If you want to jump to the end to see a full featured React setup, the Create Snowpack App React template comes with everything you'll learn in this tutorial plus other useful tools.

Reviewer 3: Perhaps a link to jump to the end of the document with a command to generate a Create Snowpack App React project?

> Prerequisites: Snowpack is a command line tool installed from npm. This guide assumes a basic understanding of Node.js, npm, and how to run commands in the terminal. Knowledge of React is not required, Snowpack is a great way to learn React.

Reviewer: okay I know we said no exclamation marks, but mannnnnnn I would add one at the end of that last sentence if I were writing that. Maybe: "Knowledge of React is not required. (Snowpack is a great way to learn React!)"?

this may just be a bad habit of mine :)

Author: We don't have to follow their standards, I'd say gonna just override here.

## Getting started with Snowpack and React

> In this step by step you'll learn how to use the [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa) tool to spin up a working Snowpack project and then how to add React to it.

Reviewer: I get what you were going for here, but it makes it sound like the goal is to learn how create-snowpack-app works. I'd rephrase to:

The easiest way to start a new Snowpack project is with [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa)). `create-snowpack-app` is a tool to quickly set up Snowpack in a new directory. For this tutorial we're going to start with the bare-bones `@snowpack/app-template-minimal` template to quickly create a new directory with some basic HTML, CSS, and JavaScript files.

Run the following commands in a new terminal window to get started:

```bash
npx create-snowpack-app my-new-project --template @snowpack/project-template-minimal
cd my-new-project
npm run start
```

> In this step by step you'll learn how to use the [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa) tool to spin up a working Snowpack project and then how to add React to it.

Reviewer 3: > [`create-snowpack-app`](https://www.snowpack.dev/#create-snowpack-app-(csa)

unbalanced parenthesis

> `create-snowpack-app` is a command line tool that creates a working Snowpack App based on a template. The template you're starting with here is `@snowpack/project-template-minimal is` a bare-bones setup where Snowpack serves an index.html, index.css, and index.js.

Reviewer 3: > `@snowpack/project-template-minimal is`

codetick ends later than it should

> `create-snowpack-app` is a command line tool that creates a working Snowpack App based on a template. The template you're starting with here is `@snowpack/project-template-minimal is` a bare-bones setup where Snowpack serves an index.html, index.css, and index.js.

Reviewer 3: > index.html, index.css, and index.js

are all not encased in backticks whilst a few lines below, they are

> This is about as basic as it gets. When you run `npm start` Snowpack serves an `index.html`, `index.css`, and `index.js`.

```bash
npx create-snowpack-project react-snowpack --template @snowpack/project-template-minimal
cd react-snowpack
npm run start

> ```

Reviewer 3: add equivalent yarn commands, which i *believe* would be

```shell
yarn create snowpack-project react-snowpack --template @snowpack/project-template-minimal
# idk if you need to cd or what not, package doesn't exist on npm so i didn't bother checking more
```

[yarn create](https://classic.yarnpkg.com/en/docs/cli/create/)

Image: screenshot of project-template-minimal, which shows "Hello world" in text on a white background

> This is about as basic as it gets. When you run `npm start` Snowpack serves an `index.html`, `index.css`, and `index.js`.

Reviewer: when you run commands like these, it's good to check in and confirm the expected result. 

> Lets walk through what this command just did: `TODO`

And then after that:

> Run `npm start` to see your new website up and running! You should see ... `TODO`

> The `README.md` contains useful information about what each file does.

> The first thing that you'll want to do is install React in your project. To do that, run the following command in your project directory:

Reviewer: I do like this in the same section, but I think it's still good to check in with the dev and signify "okay, lets move on to another thing". 

> Now that you have a basic project up and running, it's time to install React. To do that...

> To install these you can run:

Reviewer 3: > To do that, run the following command in your project directory:
>
> To install these you can run:

odd

Author: oops haha

```bash

> npm install react react-dom --save

Reviewer 3: equivalent yarn command maybe?

```shell
yarn add react react-dom
```

Reviewer: I think it's safe to focus on npm for now, since we use npm everywhere else in the guide. 

If we really wanted to, maybe somewhere in the intro / prerequisits section we could mention something like "If you prefer yarn or pnpm, fell free to use those as well"

```

> > In this tutorial commands use `--save` for frontend libraries and `--save-dev` for your developer tools to keep them organized separately. This is totally up to you, though.

Reviewer: Not convinced that this is required in our guide. Maybe keep it, but call it out as an "advanced" point. Maybe prefixed this section with "Advanced" and end with "This is totally up to you, though. Your project will run regardless of which you use".

## Create your first React component

> React uses a syntax called [JSX](https://reactjs.org/docs/introducing-jsx.html) for components. Snowpack has built in support for JSX files with the extension `.jsx`. To create your first React component in Snowpack rename `index.js` to `index.jsx`.

Reviewer: > Snowpack has built in support for JSX in files with the `.jsx` extension. That means that there's no additional plugins or configuration needed to write your first React component, but you will need to rename your `index.js` file to `index.jsx` to take advantage of this.

> > You may have used JSX in `.js` files before, but Snowpack expects JSX to be .jsx files

Reviewer: I think the comment I made above means you don't need to hit on this point twice.

```bash
mv index.js index.jsx
```

> > Remember that your index.html file is importing `/index.js`, not `/index.js`. Do you need to update it to match the new file? You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).

Reviewer 3: > is importing `/index.js`, not `/index.js`

same file name twice?

> > Remember that your index.html file is importing `/index.js`, not `/index.js`. Do you need to update it to match the new file? You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).

Reviewer 3: > You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).

add why, "which wouldn't run in the browser because it's file extension isn't .js"

> > Remember that your index.html file is importing `/index.js`, not `/index.js`. Do you need to update it to match the new file? You don't because Snowpack turns `index.jsx` (which wouldn't run in the browser) into the `index.js` (which would).

Reviewer: Could be good to phrase as a question: 

> Question: Do I need to update my index.html script tag to point to "index.jsx"?
> Answer: No! Because JSX isn't supported in the browser, Snowpack builds your file to a runnable `.js` file. This is good to remember when you're referencing built files in HTML `<script src="">` and `<link href="">` elements.

I'm not too attached to the Q&A style, but teaching the larger point here is the reason it's worth calling out vs. ignoring entirely.

> You can now import React into `index.jsx` and render a simple test component just to make sure it's working.

Reviewer: nit: "import React in `index.jsx` and render`

> Import React and add this test component `index.jsx` so you can use it.

Reviewer 3: you've already stated your intentions in the above line

> Import React and add this test component `index.jsx` so you can use it.

Reviewer: unnecceary line? duplicates what was said above

```diff
/* Add JavaScript code here! */
- console.log('Hello World! You did! Welcome to Snowpack :D');
+import React from 'react';
+import ReactDOM from 'react-dom';
+ReactDOM.render(<div>"HELLO REACT"</div>, document.getElementById('root'));

> ```

Reviewer 3: using `diff` makes it more work on the user to copy paste the code, as they have to delete the pluses. should just show the entirety of the js if possible (or figure out on the preview site how to make the +s and -s not affect copying the code)

Author: Hmm very true, any opinions here @drwpow @FredKSchott ?

Reviewer: As long as it's possible to select these lines for copy without the `+` and `-` attached, then I like this format.

Reviewer: I like the diff! As long as the `+` and `-` don't automatically come with each copy paste.

Reviewer 2: Diff or no, I think the indentation is off (click View File) here. Not an issue for now in draft, but would like to make sure the indents are aligned before publishing.

> If you run this code you'll get `Error: Target container is not a DOM element.` The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to index.html and add one in the `<body>`

Reviewer 3: several minor thigns:
- `.` in backticks for error
- `index.html` not being in code ticks despite files being in code ticks everywhere else
- comma after "if you run this code" feels natural

fixed:
```
If you run this code, you'll get `Error: Target container is not a DOM element`. The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to `index.html` and add one in the `<body>`.
```

> If you run this code you'll get `Error: Target container is not a DOM element.` The call to `document.getElementById('root')` is failing because the HTML page has no element with that ID. Head to index.html and add one in the `<body>`

Reviewer: Whenever you give instructions that fail intentionally, it's good to make it very clear. Sometimes people will even add a comment to the code snippet itself, but in this case I don't think that's warranted since the code is fine.

Something like:

> But wait! When yuo run this code in the browser, you'll see an error: "Target container is not a DOM element." This is because the call to `document.getElementById('root')` is failing because...

A screenshot of the error popup will also highlight that this is expected

Author: Hmm since the tutorial already quite long I wonder if we can just omit the error section. I'm going to try it out and see how it looks.

```diff
  <body>
-    <h1>Welcome to Snowpack!</h1>
+    <div id="root"></div>
    <script type="module" src="/index.js"></script>

> ```

Reviewer 3: see: previous comment related to `diff`

> Now it should run and you should see "HELLO REACT"

Reviewer: Section ends abruptly. Would be good to recap the main point(s?) and do some sort of celebration. You just wrote your first React component in Snowpack!!!

Image: screenshot of the project, which shows "HELLO REACT" on a white background

> ## Organizing your files

Reviewer: just a thought: I wonder if we want this headline to stick out to someone reading the guide but following along in their own application. That use would really care about "Customizing your project layout" or structure. They'd probably want to see the word "Configuring" somewhere in there as well, but we need to balance that with our whole "low config" vibe.

Maybe "Customizing your File Layout"? What do you think?

Author: I think the word layout is a bit confusing, at least to me. I think of layout as being a front-end thing?

Author: Gonna try file structure

> If you're going to be adding a bunch of new files, you probably don't want them in the root directory. A standard way of organizing React projects is an `src` directory for components and their associated files. Then a public directory for static assets like images, fonts, icons, and sometimes CSS. To do this you'll need to reorganize your files, then tell Snowpack about it in the configuration file, and update your paths in index.html.

Reviewer 3: - `To do this,`
- `index.html`

> If you're going to be adding a bunch of new files, you probably don't want them in the root directory. A standard way of organizing React projects is an `src` directory for components and their associated files. Then a public directory for static assets like images, fonts, icons, and sometimes CSS. To do this you'll need to reorganize your files, then tell Snowpack about it in the configuration file, and update your paths in index.html.

Reviewer: "If you're going to be adding a bunch of new files" sounds conditional, instead: 

> As your project grows...

Reviewer: Also lots of information contained in that last sentence, even I got lost a bit :) 

If it's too many steps to say in one sentence, maybe just hit the bigger point here? 

> Lets update our project to use this popular structure.

Reviewer: Also lets highlight that we don't lock you into a single cookie cutter layout: Full rewrite of this paragraph:

> If you're going to be adding a bunch of new files, you probably don't want them in the root directory. Snowpack is flexible enough to support whatever project layout that you prefer. For this guide, let's use a popular project layout from the React communtiy.
> 
> A standard way of organizing React projects is an `src` directory for components and their associated files. Then a `public` directory for static assets like images, fonts, icons, and sometimes CSS. Lets update our project to use this popular structure.

```
📁 src
↳ index.jsx
📁 public
↳ index.css
↳ index.html
```

> 

Reviewer 3: maybe there should be some text here that says what the commands are doing? "Provided are some commands that give the above project structure:"

> ```bash

Reviewer: maybe add a comment: If it feels more familiar, feel free to use your computer's visual file browser as well.

mkdir src
mkdir public
mv index.jsx src/index.jsx
mv index.html public/index.html
mv index.css public/index.css
```

> You need to tell Snowpack about this new structure using the [Snowpack Configuration API](https://www.snowpack.dev/#configuration) in the `snowpack.config.js` file. Right now it's a bare bones scaffold that doesn't do anything. Add this to the `mount` object:

Reviewer: Good idea to mention the API, but not so fast. As a reader, I'd click it immediately and then get overwhelmed. How about:

Once your new structure is in place, you need to tell Snowpack about it. Every Snowpack project comes with a `snowpack.config.js` file for any configuration that you might need. Right now, you should see a mostly-empty, bare-bones config file. Add this to the empty `mount` object:

```
{CODE SNIPPET}
```

In the section below, you can link to the full API reference

```diff
  mount: {
-    /* ... */
+    public: '/',
+    src: '/_dist_',
  },
```

> This tells it look at `public` instead of look at `/` for the index. For example when building `index.css`, Snowpack looks for this files in `./public/index.css`.

Reviewer 3: `this files`

> This tells it look at `public` instead of look at `/` for the index. For example when building `index.css`, Snowpack looks for this files in `./public/index.css`.

Reviewer: start with a single sentence explaining what "mount" does in general.

Author: I made an attempt but LMK how it is

> And it also tells it to put the final output files in a subdirectory called `_dist_` so change that path in `index.html` from `/index.js` to `/_dist_/index.js`.

Reviewer 3: add a comma after "_dist_`, so change that", was surprising that the sentence turned into a command midway reading

```diff
    <h1>Welcome to Snowpack!</h1>
    <div id="root"></div>
-    <script type="module" src="/index.js"></>
+    <script type="module" src="/_dist_/index.js"></script>
  </body>
```

If Snowpack is still running, exit the process, you need to restart Snowpack for configuration file changes. When you start up again, if it worked, it should look the same.

> Create a new file at `src/App.jsx` and paste the following code into this new file to create an `App` component:

Reviewer: I'm leaning towards this `src/App.jsx` section being its own, short section. Because we outline the structure that we want to move to at the top, it feels weird to end this section with a different set of files than we started.

```jsx
import React, {useState, useEffect} from 'react';

function App() {
  // Create the count state.
  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);
  }, [count, setCount]);

>   // Return the App component.

Reviewer 3: someone reading this probably already knows react. since we're in a guide, should try to keep the amount of code minimal.

Reviewer: A stated goal here was to get you something that matched the current `create-snowpack-app` React template. If you see that template, you'll see it's basically an exact match to what you end up with here.

We also need the counter to highlight HMR & Fast Reload in the section below/

I'm personally not against the extra lines of copy-paste here, but @melissamcewen it could be worth revisiting/confirming this goal. Do we need to match the create-snowpack-app react template EXACTLY, or can we just match it in spirit? If just in spriit, then I'd agree we could probably trim some CSS, or include a Snowpack logo in the `minimal` template and then use that here instead of the React logo.

  return (
    <div className="App">
      <header className="App-header">
        <p>
          Edit <code>src/App.jsx</code> and save to reload.
        </p>
        <p>
          Page has been open for <code>{count}</code> seconds.
        </p>
        <p>
          <a
            className="App-link"
            href="https://reactjs.org"
            target="_blank"
            rel="noopener noreferrer"
          >
            Learn React
          </a>
        </p>
      </header>

>     </div>

Reviewer 3: this takes up a lot of lines in a guide - can it be trimmed?

  );
}

export default App;
```

> Now include it in `index.js`

Reviewer 3: `index.jsx`

```diff
import React from 'react';
import ReactDOM from 'react-dom';

> + import App from './App';

Reviewer 3: isn't it `App.jsx` or does snowpack do magic here?

Author: I'm pretty sure Snowpack does magic. But @FredKSchott would know more.

Reviewer: Both are supported, but its best to get users in the habit of including the file extension

> + import App from './App';

Reviewer: lets do './App.jsx' here. We want people to get into that habit.

- ReactDOM.render(<div>"HELLO WORLD"</div>, document.getElementById('root'));
+ ReactDOM.render(
+    <React.StrictMode>
+      <App />
+    </React.StrictMode>,
+    document.getElementById('root'),
+  );

> ```

Reviewer 3: so much is being added, can probably not do the `diff` thing

Reviewer: I do like the diff, but the mixing of add/remove does make copy-paste harder. Can we move the 1 red line to above the `import App from './App';` line, so that all new lines are together?

> [Strict Mode](https://reactjs.org/docs/strict-mode.html) is a tool for highlighting potential problems in React code.

You shouldn't need to restart Snowpack to see this, it should look like this:

> IMAGE: screenshot of the project, showing text saying "edit `src/App.js` and save to reload," text that says "Page has been open for" and the number of seconds then "seconds". A link with the text Learn React that links to the React website.

Reviewer 3: make sure the image says "edit `src/App.jsx` ..."

## Styling your Snowpack/React project

When you add assets like images or CSS, Snowpack includes them in your final build. If you already know React, this process should look pretty familiar.

> > As you're doing this, you should not need to reload the page or restart Snowpack. Snowpack updates the project in the browser as you edit code, which is pretty nifty.

Reviewer: Snowpack automatically updates

> Add this file `logo.svg` (link to file) to `src`. Now you can import it into your `App.jsx` and use it in an `img` tag to display it.

Reviewer 3: shouldn't assets like `logo.svg` be in `public`? earlier in the guide it said

> Then a public directory for static assets like images, fonts, icons, and sometimes CSS.

Author: Good question, I'll leave this one open. I think sometimes if assets are component specific people will add them in the component directory.

Reviewer: +1 for @melissamcewen's answer here, although that's a good point of possible confusion. Maybe worth clarifying with an added line to this section like "for some static assets, you may want to locate them near the React components that use them..."

and clarifying the public line with something like "a public directory for **site** assets like images, fonts, icons, and sometimes CSS."

> Add this file `logo.svg` (link to file) to `src`. Now you can import it into your `App.jsx` and use it in an `img` tag to display it.

Reviewer: nit: to your `src/` directory.

```diff
import React, { useState, useEffect } from 'react';
+ import logo from './logo.svg';

function App() {
  // Create the count state.
  const [count, setCount] = useState(0);
  // Create the counter (+1 every second).
  useEffect(() => {
    const timer = setTimeout(() => setCount(count + 1), 1000);
    return () => clearTimeout(timer);
  }, [count, setCount]);
  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">
+        <img src={logo} className="App-logo" alt="logo" />
```

Image: the React logo (a blue atom) is now at the top of the page

> The project already has index.css for global styles, but component CSS can help keep code organized. Snowpack knows to only load this CSS when displaying the `App` component, which means it's not loading unnecessary CSS.

Reviewer 3: "component CSS" how about "having CSS for each component"? unless "component CSS" is an official react term, then use that cuz it's shorter

Create `src/App.css` and add this CSS:

```css
.App {
  text-align: center;
}
.App code {
  background: #fff3;
  padding: 4px 8px;
  border-radius: 4px;
}
.App p {
  margin: 0.4rem;
}
.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-spin infinite 20s linear;
  }
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }

> }

Reviewer 3: what's the bare minimum to get the logo to spin and things to be centered okay-ishly, this is a lot of scrolling lol

```

To use it, import it `App.jsx` with

```diff
import logo from './logo.svg';
+import './App.css';
```

Image: screenshot or GIF. The page now has centered items, a grey background, styled fonts, and the React logo has an animation that rotates it

You may notice the page has a white border, which doesn't look so great. That's due to a global style. Go to `index.css` and edit the CSS to remove it:

```diff
/* Add CSS styles here! */
body {
  font-family: sans-serif;
-  margin: 2em 1em;

> +  margin: 0;

Reviewer: hahaha good catch. lets make this the default in the minimal template so that we don't need this extra step.

  line-height: 1.5em;
}
```

> Image:

Reviewer 3: not sure if this being left blank or not is intentional so im commenting, but easily inferrable so /shrug

> ## Adding React Fast Refresh with a Snowpack plugin

Reviewer: How about "Making Snowpack Even Faster with Fast Refresh"

or "Developing Even Faster with Fast Refresh"

> [React Fast Refresh](https://reactnative.dev/docs/fast-refresh)? What's that? It's a developer tool that makes it much easier to see how updates to your code work in the browser. React projects are often interactive and include state. For example, this project you're building has a state that is the amount of time on the page. When developing with state it's useful not to lose it while you edit code. React Fast Refresh shows you updates without refreshing the entire page. Showing you how to add this is also a good intro to Snowpack plugins. Snowpack starts with a minimal setup with the perspective that you can add what you need through the plugin system.

Reviewer: Running out of time, but would love to take a pass on this section later. Overall I like this new approach though.

Reviewer 2: Love the conversational tone here

Start by enabling [Hot Module Replacement](https://www.snowpack.dev/#hot-module-replacement), a requirement for Fast Refresh. Snowpack automatically refreshes to show changes in your code. Hot Module Replacement only updates the components that changed without refreshing the whole page. You can enable it for React with a small snippet of code in `index.jsx`.

Head to `index.jsx` and add this snippet at the bottom:

```diff
ReactDOM.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
  document.getElementById('root'),
);
+// Hot Module Replacement (HMR) - Remove this snippet to remove HMR.
+// Learn more: https://www.snowpack.dev/#hot-module-replacement
+if (import.meta.hot) {
+  import.meta.hot.accept();

> + }

Reviewer 3: `+ }` -> `+}` for consistency

```

Now when you change `App.jsx` the changes show without the whole page refreshing.

Image: GIF showing code side by side with the project. A change in made to App.jsx and it shows when the changed file is saved. But the counter resets to 0.

However, if you make a change to App.jsx, you'll see the counter on the page reset to 0, even though the page itself hasn't fully refreshed. React Fast Refresh solves this.

Installing a plugin involves installing the module and then enabling it in the Snowpack configuration file.

Install the module first by running.

```bash

> npm install @snowpack/plugin-react-refresh --save-dev

Reviewer 3: yarn eq: `yarn add @snowpack/plugin-react-refresh --dev`

```

To enable it head to `snowpack-config.js` and add it to the array of plugins

```diff
module.exports = {
  mount: {
    public: '/',
    src: '/_dist_',
  },
  plugins: [
+    '@snowpack/plugin-react-refresh'
  ],
};
```

> > you'll need to restart Snowpack

Reviewer 3: "... for the configuration changes to take effect"

Now make an edit to `App.jsx`. If this worked you'll see it without the counter stopping or losing its value.

Image: GIF showing code side by side with the app. A change in made to App.jsx and it shows immediately when the file is changed. The counter keeps counting uninterrupted.

## Going further

Great job! You're now ready to build the React project if your dreams with Snowpack.

Image: Certificate with share buttons for Twitter

At this point you the basics and have a great starter for any React project. But if you take a look at our Snowpack React template you'll notice it has some other tools you might find useful

- [Prettier](https://prettier.io/) – a code formatter

> - Tests

Reviewer 3: mention mocha/jest? (similar to what was done above for prettier)

    return () => clearTimeout(timer);
  }, [count, setCount]);
  // Return the App component.
  return (
    <div className="App">
      <header className="App-header">
+        <img src={logo} className="App-logo" alt="logo" />
  function App() {
    // Create the count state.
    const [count, setCount] = useState(0);
    // Create the counter (+1 every second).
    useEffect(() => {
      const timer = setTimeout(() => setCount(count + 1), 1000);
      return () => clearTimeout(timer);
    }, [count, setCount]);
    // Return the App component.
    return (
      <div className="App">
        <header className="App-header">
+       <img src={logo} className="App-logo" alt="logo" />
        <p>
```

<div class="frame"><img src="/img/guides/react/minimalist-hello-world-react-logo.png" alt="the React logo (a blue atom) is now at the top of the page" class="screenshot"/></div>

The project already has index.css for global styles. For CSS that's only for a specific component, a common design pattern is to add it in a CSS file with the same base name as the component. The style file for `App.jsx` would be `App.css` with this pattern.

> > 💡 Tip: Snowpack has built-in support for [CSS Modules](https://www.snowpack.dev/#import-css-modules) and if you'd like to use SASS there is an official [SASS Plugin](https://www.snowpack.dev/#sass).

Reviewer 2: ```suggestion
> 💡 Tip: Snowpack has built-in support for [CSS Modules](https://www.snowpack.dev/#import-css-modules) and if you'd like to use Sass there is an official [Sass Plugin](https://www.snowpack.dev/#sass).
```
